### PR TITLE
[INTEL MKL] Add Weight caching in MKL Quantized Matmul Op

### PR DESCRIPTION
**Pull request by rgomathi**
This PR adds weight buffer caching that improves the performance in scenarios where the weight buffer format is not same as the MKL-DNN expected format. This is achieved by reordering the weight for the first time and caching the same for re-using it in subsequent iterations.

### CODE REVIEW

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 1]
```diff
@@ -434,8 +461,90 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Toutput> {
...
+  Tweight* GetCachedWeight(OpKernelContext* context,
+                           const memory::format& weight_mf)
+      LOCKS_EXCLUDED(mu_) {
+    tf_shared_lock lock(mu_);
+    const Tensor& weight_t = *weight_oi.AccessTensor(context);
+    const Tensor& weight_md_t = *weight_oi_md.AccessTensor(context);
+
+    // Check if the  memory descriptor of the cached weight is same as
+    // weight_mf. If so use the cached memory, else return NULL
+    if (weight_md_t.scalar<int32>().size() &&
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op.cc]: Nit: Let's make this check explicit.
```suggestion
    if (weight_md_t.scalar<int32>().size() > 0 &&
```

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 2]
```diff
@@ -434,8 +461,90 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Toutput> {
...
+
+    void* weight_oi_t_data = weight.GetTensorBuffer(weight_tensor_ptr);
+    size_t weight_size = weight.GetOpMem().get_primitive_desc().get_size();
+    memcpy(weight_oi_t_data, weight_data, weight_size);
+
+    // Cache the memory descriptor
+    Tensor* weight_md_tensor_ptr = nullptr;
+    TensorShape weight_mkl_format;
+    weight_mkl_format.AddDim(
+        sizeof(
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op.cc]: Why can't we just use 1? Later on you read the value using `scalar`. This means the size must always be 1. 

If you want to be flexible, then I think you can replace `sizeof(matmul_fwd_pd.get()...)` with `sizeof(memory::format)`. And you'll need to read the values as vector and do a vector comparison. (The two places should have consistent assumptions.)

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 3]
```diff
@@ -434,8 +461,90 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Toutput> {
...
+
+  Tweight* GetCachedWeight(OpKernelContext* context,
+                           const memory::format& weight_mf)
+      LOCKS_EXCLUDED(mu_) {
+    tf_shared_lock lock(mu_);
+    const Tensor& weight_t = *weight_oi.AccessTensor(context);
+    const Tensor& weight_md_t = *weight_oi_md.AccessTensor(context);
+
+    // Check if the  memory descriptor of the cached weight is same as
+    // weight_mf. If so use the cached memory, else return NULL
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op.cc]: Can the weight format change if the weight is constant? I'm not sure we need to check this.

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 4]
```diff
@@ -471,6 +471,92 @@ TEST_F(QuantizedMatMulTest, Small_withBiasAndReluAndReq) {
   test::ExpectTensorEqual<quint8>(expected, output_quantized);
 }
 
+// Two small matrices A of type uint8 and B of type int8  are multiplied
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Nit:
```suggestion
// Two small matrices A of type uint8 and B of type int8 are multiplied
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 5]
```diff
@@ -471,6 +471,92 @@ TEST_F(QuantizedMatMulTest, Small_withBiasAndReluAndReq) {
...
+  AddInputFromArray<uint8>(kDummyShape, kDummyTensor);
+
+  TF_ASSERT_OK(RunOpKernel());
+  // Here are the results we expect, from hand calculations:
+  // (1 * 7) + (2 * 11) + (3 * 15) = 74
+  // (1 * 8) + (2 * 12) + (3 * 16) = 80
+  // (1 * 9) + (2 * 13) + (3 * 17) = 86
+  // (1 * 10) + (2 * 14) + (3 * 18) = 92
+  // Final result after Bias addition:
+  // 74  + 1 = 75 , 80  + 2 = 82 , 86  + 3 = 89 , 92  + 4 = 96,
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Thank you for the detailed comments! :)

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 6]
```diff
@@ -471,6 +471,92 @@ TEST_F(QuantizedMatMulTest, Small_withBiasAndReluAndReq) {
...
+  Tensor output_quantized;
+  conv_comp.ConvertMKL2TF<qint32>(DT_QINT32, output, mkl_shape_tensor,
+                                  output_quantized);
+
+  test::ExpectTensorEqual<qint32>(expected, output_quantized);
+
+  // Test for the second time to use the cached weight
+  // Though we rely on performance as well as mkldnn verbose to
+  // determine if the weights buffer is cached and reused.
+  TF_ASSERT_OK(RunOpKernel());
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Can we time both calls and check (here) that the second call is faster?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR! I have a few high-level questions/comments.

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 7]
```diff
@@ -244,13 +250,34 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Toutput> {
...
+        bool is_weight_cached = false;
+        // For Batch size 1, MKLDNN expected weight format is OI whereas
+        // TF default format is IO. So in that case convert weight from IO
+        // to OI for the first iteration and cache it to reuse in the
+        // subsequent iterations, if the weight is constant.
+        if (is_weight_const_) {
+          // Check if the weight is already cached or not
+          if (IsWeightCacheEmpty(context)) {
+            // Cache weight if it is not cached.
+            CacheWeight(context, matmul_fwd_pd, weight_data, weight_tensor,
```
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: Please ignore this comment. I had written it before I read `CacheWeight` and I forgot to delete it.

**Comment by gbaned** [comment]
@rgomathi Could you please check reviewer comments and keep us posted. Thanks!

**Comment by rgomathi** [comment]
Hi @gbaned, I addressed the review comments and just double checking them with colleagues internally. I will submit them ASAP. Sorry for the delay.

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 8]
```diff
@@ -4,7 +4,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+http://www.apache.org/licenses/LICENSE-2.0
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: changed it back..

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 9]
```diff
@@ -244,13 +250,34 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Toutput> {
...
+
       Tweight* weight_data = nullptr;
       if (weight_md.data.format != matmul_fwd->GetweightMemoryFormat()) {
-        weight.SetUsrMem(weight_md, &weight_tensor);
-        weight.CheckReorderToOpMem(
-            matmul_fwd_pd.get()->weights_primitive_desc());
-        weight_data =
-            static_cast<Tweight*>(weight.GetOpMem().get_data_handle());
+        bool is_weight_cached = false;
+        // For Batch size 1, MKLDNN expected weight format is OI whereas
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: changed as commented...

File: tensorflow/core/kernels/mkl_qmatmul_op.cc [hunk 10]
```diff
@@ -434,8 +461,90 @@ class MklDnnQuantizedMatMulOp : public MklDnnMatMulOpBase<Toutput> {
...
 
   int mode_;
+  bool is_weight_const_;
+  // LOCKS_EXCLUDED annotation ensures that the lock (mu_) cannot
+  // be acquired before entering the function, since it is acquired
+  // inside the function.
+  inline bool IsWeightCacheEmpty(OpKernelContext* context) LOCKS_EXCLUDED(mu_) {
+    tf_shared_lock lock(mu_);
+    const Tensor& weight_t = *weight_oi.AccessTensor(context);
+    return (weight_t.NumElements() == 0);
```
[the remark this answers is not in the document]
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: changed accordingly..

File: tensorflow/core/kernels/mkl_qmatmul_op.cc (hunk 1 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: Done

File: tensorflow/core/kernels/mkl_qmatmul_op.cc (hunk 2 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: Changed to just 1 as it makes sense for now, thanks for your suggestion !

File: tensorflow/core/kernels/mkl_qmatmul_op.cc (hunk 3 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: Yes, MKL-DNN expects different format if the batch size changes. For the same weight the expected format is IO for higher BS and OI for BS=1. Hope this makes sense.

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc (hunk 4 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Done!

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc (hunk 5 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Thanks :)

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc (hunk 6 above)
Comment (rgomathi) [reply tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Added code as commented.

**Comment by rgomathi** [comment]
Hi @penpornk,  I've addressed the review comments and pushed the patch. Please review. Sorry for the delay....

File: tensorflow/core/kernels/mkl_qmatmul_op.cc (hunk 3 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl_qmatmul_op.cc]: Got it. Thank you for the clarification!

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc (hunk 6 above)
Comment (penpornk) [reply tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Thank you very much!

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 11]
```diff
@@ -471,6 +471,102 @@ TEST_F(QuantizedMatMulTest, Small_withBiasAndReluAndReq) {
...
+
+  // Test for the second time to use the cached weight
+  start_time = Env::Default()->NowMicros();
+  TF_ASSERT_OK(RunOpKernel());
+  end_time = Env::Default()->NowMicros();
+  int64 total_duration_opt = end_time - start_time;
+  LOG(INFO) << " Time taken by first call : " << total_duration_unopt
+            << ", Time taken after Caching : " << total_duration_opt;
+
+  // Cached call should be atleast 20% faster
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Nit: 
```suggestion
  // Cached call should be at least 20% faster.
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the fixes! One minor change please.

**Comment by rgomathi** [comment]
Hi @penpornk , I accepted the change and committed here itself . Hope it is fine. Thanks

**Comment by rgomathi** [comment]
Hi @penpornk , Please let me know if anything else is pending from my side. Thanks

**Review by penpornk — APPROVED** [review]
So sorry for the delay! Thank you for the change!

**Comment by rgomathi** [comment]
No worries and Thanks !

**Pull request merged**